### PR TITLE
NIFI-1583 Added ConvertJSONtoCSV processor

### PR DESCRIPTION
**Pull request by acesir**
New functionality to convert complicated multi nested JSON format into
tabular 2 dimensional CSV format

### CODE REVIEW

**Comment by acesir** [comment]
Removing the Atomic reference makes sense. 

As far as removing the fields goes, yes, both would be removed. Code could be updated where siblings.name needs to be explicitly defined in order to remove the nested name field. Currently the desire was to remove fields at every level to make the functionality itself simpler.

**Comment by joewitt** [comment]
I am not sure we can use org.json:json dependency.  Its license as found in the readme of https://github.com/stleary/JSON-java/blob/master/README says  "The license includes this restriction: "The software shall be used for good,
not evil." If your conscience cannot live with that, then choose a different
package.""

I'm not sure this is acceptable in that it doesn't claim to be public domain or be part of any approved ASF friendly license.

**Comment by mattyb149** [comment]
I think the ASF is ok with the JSON license: http://www.apache.org/legal/resolved.html#json
@joewitt do you agree?

**Comment by mattyb149** [comment]
@acesir Mind rebasing against the latest master? Please and thanks!

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml [hunk 1]
```diff
@@ -221,6 +221,11 @@ language governing permissions and limitations under the License. -->
             <version>1.4.187</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: the standard bundle has jackson dependencies already, does this need org.json.json to accomplish the task?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    volatile String emptyFields = "";
+
+    public static final AllowableValue INCLUDE_HEADER_TRUE = new AllowableValue(
+            "True", "True", "Creates headers for each JSON file.");
+    public static final AllowableValue INCLUDE_HEADER_FALSE = new AllowableValue(
+            "False", "False", "Only parses the JSON fields and does not include headers");
+
+
+    public static final PropertyDescriptor DELIMITER = new PropertyDescriptor
+            .Builder().name("CSV Delimiter")
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java]: The latest convention is to include a name() that is language independent, with displayName() that can be more user-focusing. This is just a suggestion that will help with internationalization later.

Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java]: Also maybe just call it Delimiter. I know the processor is called ConvertJSONtoCSV but a tab-delimited file is technically a TSV :-/

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    public static final PropertyDescriptor EMPTY_FIELDS = new PropertyDescriptor
+            .Builder().name("Empty field value")
+            .description("During denormalization/flattening of the JSON the value that will be substituted for empty fields values " +
+                    "(Example: NULL). Defaults to empty string if not specified.")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor INCLUDE_HEADERS = new PropertyDescriptor
+            .Builder().name("Include Headers")
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java]: Should there be a "Number of header lines" property like there is in SplitText?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    public static final Relationship RELATIONSHIP_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("Successfully parsing the JSON file to CSV ")
+            .build();
+
+    public static final Relationship RELATIONSHIP_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("Failed parsing the JSON file to CSV ")
+            .build();
+
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java]: Should there be an "original" relationship that the JSON file would be transferred to upon success? Other processors do similar things.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+@Tags({"JSON, CSV, convert"})
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java]: seems like tags are all lowercase in the processors I've seen, not sure this is required or not

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+
+            return flatJson;
+        }
+
+        private static List<Map<String, String>> handleAsArray(String json) throws JSONException {
+            List<Map<String, String>> flatJson = null;
+
+            try {
+                JSONArray jsonArray = new JSONArray(json);
+                flatJson = parse(jsonArray);
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java]: Can this be parallelized somehow? Fork/Join or whatever? Not sure about overhead, just asking :)

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+            }
+
+            for (Map<String, String> map : flatJson) {
+                csvData.append(getSeperatedColumns(headers, map, separator, emptyFields) + "\n");
+            }
+
+            return csvData;
+        }
+
+        private static String getSeperatedColumns(Set<String> headers, Map<String, String> map, String separator, String emptyFields) {
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ConvertJSONtoCSV.java]: Seperated should be Separated, sorry minor typo but while you're in there :)

**Comment by trixpan** [comment]
@acesir wondering if you had time to rebase the PR so we can complete review and possibly merge it?

**Comment by joewitt** [comment]
I plan to close out this PR in a stale PR sweep soon.

**Pull request closed without merging**